src/pkg/cli/login_test.go

```diff
@@ -0,0 +1,154 @@
+package cli
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+
+	"github.com/DefangLabs/defang/src/pkg/cli/client"
+	defangv1 "github.com/DefangLabs/defang/src/protos/io/defang/v1"
+)
+
+func TestGetExistingToken(t *testing.T) {
+	fabric := "test.defang.dev"
+
+	t.Run("Get access token from environmental variable", func(t *testing.T) {
+		expectedToken := "env-token"
+		currentToken := os.Getenv("DEFANG_ACCESS_TOKEN")
+		if currentToken != expectedToken {
+			os.Setenv("DEFANG_ACCESS_TOKEN", expectedToken)
+		}
+		defer os.Setenv("DEFANG_ACCESS_TOKEN", currentToken)
+
+		accessToken := GetExistingToken(fabric)
+		if accessToken != expectedToken {
+			t.Errorf("expected %s, got %s", expectedToken, accessToken)
+		}
+	})
+
+	t.Run("Get access token from file", func(t *testing.T) {
+		expectedToken := "file-token"
+		tokenFile := getTokenFile(fabric)
+		os.WriteFile(tokenFile, []byte(expectedToken), 0600)
+
+		currentToken := os.Getenv("DEFANG_ACCESS_TOKEN")
+		os.Unsetenv("DEFANG_ACCESS_TOKEN")
+
+		defer func() {
+			os.Remove(tokenFile)
+			os.Setenv("DEFANG_ACCESS_TOKEN", currentToken)
+		}()
+
+		accessToken := GetExistingToken(fabric)
+		if accessToken != expectedToken {
+			t.Errorf("expected %s, got %s", expectedToken, accessToken)
+		}
+	})
+}
+
+type MockGitHubAuthService struct {
+	GitHubAuthService
+	MockLogin func(ctx context.Context, client client.FabricClient, gitHubClientId, fabric string) (string, error)
+}
+
+func (g MockGitHubAuthService) login(
+	ctx context.Context, client client.FabricClient, gitHubClientId, fabric string,
+) (string, error) {
+	return g.MockLogin(ctx, client, gitHubClientId, fabric)
+}
+
+func TestInteractiveLogin(t *testing.T) {
+	temp := githubAuthService
+	accessToken := "test-token"
+	fabric := "test.defang.dev"
+	tokenFile := getTokenFile(fabric)
```

Comment:
It would be great if this test didn't mutate the developer's token file locally. Lets try to set the `XDG_STATE_HOME` env var so that we use a temp directory for this test.

```suggestion
	// use a temp dir for the token file
	t.setEnv("XDG_STATE_HOME", t.TempDir())
	tokenFile := getTokenFile(fabric)
```
I haven't tested this, so let me know if you have issues

Reply:
@jordanstephens 
Thanks for sharing the beautiful way!
But I'm testing this env but it does not work as we expected. When I take a look here with debugger, these lines were executed before setting `XDG_STATE_HOME ` in my test. 

https://github.com/DefangLabs/defang/blob/main/src/pkg/cli/client/state.go#L14-L29

I will look for a way to set the env before initializing `stateDir`.

Reply:
Ah I see. I wonder if it's worth refactoring that code so that the values are lazily computed rather than constant globals.